client/app/components/dashboards/AddWidgetDialog.jsx

```diff
@@ -0,0 +1,448 @@
+  };
+
+  static propTypes = {
+    dashboard: PropTypes.object, // eslint-disable-line react/forbid-prop-types
+    close: PropTypes.func,
+    dismiss: PropTypes.func,
+  };
+
+  static defaultProps = {
+    dashboard: null,
+    close: () => {},
+    dismiss: () => {},
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      saveInProgress: false,
+      selectedQuery: null,
+      searchTerm: '',
+      recentQueries: [],
+      searchedQueries: [],
+      selectedVis: null,
+      parameterMappings: [],
+    };
+
+    // Don't show draft (unpublished) queries
+    const Query = this.props.Query; // eslint-disable-line react/prop-types
+    Query.recent().$promise.then((items) => {
+      this.setState({
+        recentQueries: items.filter(item => !item.is_draft),
+      });
+    });
+
+    const searchQueries = debounce(this.searchQueries.bind(this), 200);
+    this.onSearchTermChanged = (event) => {
+      const searchTerm = event.target.value;
+      this.setState({ searchTerm });
+      searchQueries(searchTerm);
+    };
+  }
+
+  selectQuery(queryId) {
+    // Clear previously selected query (if any)
+    this.setState({
+      selectedQuery: null,
+      selectedVis: null,
+      parameterMappings: [],
+    });
+
+    if (queryId) {
+      const Query = this.props.Query; // eslint-disable-line react/prop-types
+      Query.get({ id: queryId }, (query) => {
+        if (query) {
+          const MappingType = AddWidgetDialog.MappingType;
+          const existingParamNames = map(
+            this.props.dashboard.getParametersDefs(),
+            param => param.name,
+          );
+          this.setState({
+            selectedQuery: query,
+            parameterMappings: map(query.getParametersDefs(), param => ({
+              name: param.name,
+              type: includes(existingParamNames, param.name)
+                ? MappingType.DashboardMapToExisting : MappingType.DashboardAddNew,
+              mapTo: param.name,
+              value: param.normalizedValue,
+              param,
+            })),
+          });
+          if (query.visualizations.length) {
+            this.setState({ selectedVis: query.visualizations[0] });
+          }
+        }
+      });
+    }
+  }
+
+  searchQueries(term) {
+    if (!term || term.length === 0) {
+      this.setState({ searchedQueries: [] });
+      return;
+    }
+
+    const Query = this.props.Query; // eslint-disable-line react/prop-types
+    Query.query({ q: term }, (results) => {
+      // If user will type too quick - it's possible that there will be
+      // several requests running simultaneously. So we need to check
+      // which results are matching current search term and ignore
+      // outdated results.
+      if (this.state.searchTerm === term) {
+        this.setState({ searchedQueries: results.results });
+      }
+    });
+  }
+
+  selectVisualization(query, visualizationId) {
+    visualizationId = '' + visualizationId;
```

Comment:
In what cases it gets cast to string?

Reply:
Another benefit of using Ant's `Select` is that it preserves the original value of the input.